### PR TITLE
Feature/2064 attrs blacklist subs

### PR DESCRIPTION
**Pull request by crbrox**
@fgalan @kzangeli

### CODE REVIEW

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 1]
```diff
@@ -1390,7 +1394,7 @@ static bool processOnChangeConditionForUpdateContext
   {
     ContextAttribute* caP = notifyCerP->contextElement.contextAttributeVector[ix];
 
-    if (attrL.size() == 0)
+    if (attrL.size() == 0 || blacklist)
```
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Style-guide: parenthesis around comparison:

    if ((attrL.size() == 0) || blacklist)

[ In my opinion it's better like this:

    if ((attrL.size() == 0) || (blacklist == true))

but that's not in our style guide ]

File: src/lib/ngsi/ContextAttributeVector.cpp [hunk 2]
```diff
@@ -193,6 +193,16 @@ std::string ContextAttributeVector::toJson(bool isLastElement, RenderFormat rend
       }
     }
   }
+  else // attrsFilter is black list
+  {
+    for (unsigned ix = 0; ix < vec.size(); ++ix)
+    {
+      if ( std::find(attrsFilter.begin(), attrsFilter.end(), vec[ix]->name) == attrsFilter.end() )
```
Comment (kzangeli) [inline src/lib/ngsi/ContextAttributeVector.cpp]: Style-guide: no spaces after '(' nor before ')'

File: src/lib/ngsi/ContextAttributeVector.cpp [hunk 3]
```diff
@@ -241,7 +251,17 @@ std::string ContextAttributeVector::toJson(bool isLastElement, RenderFormat rend
       }
     }
   }
-
+  else // attrsFilter is black list
+  {
+    for (unsigned ix = 0; ix < vec.size(); ++ix)
+    {
+      if ( std::find(attrsFilter.begin(), attrsFilter.end(), vec[ix]->name) == attrsFilter.end() )
```
Comment (kzangeli) [inline src/lib/ngsi/ContextAttributeVector.cpp]: Style-guide: no spaces after '(' nor before ')'

File: src/lib/ngsi/ContextAttributeVector.h [hunk 4]
```diff
@@ -71,7 +71,8 @@ typedef struct ContextAttributeVector
 
   std::string        toJson(bool                             isLastElement,
                             RenderFormat                     renderFormat,
-                            const std::vector<std::string>&  attrsFilter);
+                            const std::vector<std::string>&  attrsFilter,
+                            bool                             blacklist = false );
```
Comment (kzangeli) [inline src/lib/ngsi/ContextAttributeVector.h]: same here

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 1 above)
Comment (crbrox) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: fixed 80d6d1455a547150d87293bcc2778de23b12df94

File: src/lib/ngsi/ContextAttributeVector.cpp (hunk 2 above)
Comment (crbrox) [inline src/lib/ngsi/ContextAttributeVector.cpp]: fixed 80d6d1455a547150d87293bcc2778de23b12df94

File: src/lib/ngsi/ContextAttributeVector.cpp (hunk 3 above)
Comment (crbrox) [inline src/lib/ngsi/ContextAttributeVector.cpp]: fixed 80d6d1455a547150d87293bcc2778de23b12df94

File: src/lib/ngsi/ContextAttributeVector.h (hunk 4 above)
Comment (crbrox) [inline src/lib/ngsi/ContextAttributeVector.h]: fixed 80d6d1455a547150d87293bcc2778de23b12df94

File: src/lib/mongoBackend/MongoCommonSubscription.cpp [hunk 5]
```diff
@@ -359,3 +360,15 @@ void setFormat(const Subscription& sub, BSONObjBuilder* b)
...
+/* ****************************************************************************
+*
+* setBlacklist -
+*
+*/
+void setBlacklist(const Subscription& sub, BSONObjBuilder* b)
+{
+  bool bl = sub.notification.blacklist;
+  b->append(CSUB_BLACKLIST, bl);
+  LM_T(LmtMongo, ("Subscription blacklist: %s", bl?"TRUE":"FALSE"));
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonSubscription.cpp]: I think in other places (eg. extended) we are using "true" and "false" (lowercase) so maybe it is better that way, to align.

In addition, I think that whitespace would be added for `?` and `:` (not sure 100%).

File: src/lib/mongoBackend/mongoUpdateSubscription.cpp [hunk 6]
```diff
@@ -499,7 +500,24 @@ static void setFormat(const SubscriptionUpdate& subUp, const BSONObj& subOrig, B
...
+{
+  if (subUp.blacklistProvided)
+  {
+    setBlacklist(subUp, b);
+  }
+  else
+  {
+    bool bl = getBoolFieldF(subOrig, CSUB_BLACKLIST);
+    b->append(CSUB_BLACKLIST, bl);
+    LM_T(LmtMongo, ("Subscription blacklist: %s", bl? "TRUE" : "FALSE"));
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoUpdateSubscription.cpp]: Same comment about TRUE/FALSE -> true/false

File: src/lib/mongoBackend/mongoUpdateSubscription.cpp [hunk 7]
```diff
@@ -499,7 +500,24 @@ static void setFormat(const SubscriptionUpdate& subUp, const BSONObj& subOrig, B
...
+*/
+static void setBlacklist(const SubscriptionUpdate& subUp, const BSONObj& subOrig, BSONObjBuilder* b)
+{
+  if (subUp.blacklistProvided)
+  {
+    setBlacklist(subUp, b);
+  }
+  else
+  {
+    bool bl = getBoolFieldF(subOrig, CSUB_BLACKLIST);
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoUpdateSubscription.cpp]: I'd suggest to use:

```
bool b1 = subOrig.hasField(CSUB_BLACKLIST)? getBoolFieldF(subOrig, CSUB_BLACKLIST) : false;
```

File: test/functionalTest/cases/2064_attrs_blacklist_subs/2064_attrs_blacklist_complex.test [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+#
+WAIT_TIME=1.5s
+#
+# 01. Create subscription
+# 02. Create entity (triggers notification)
+# 03. Check subscription timesSent value (1)
+# 04. Subcription update not involving actual update
+# 05. Update entity (triggers notification)
+# 06. Check subscription timesSent value (2)
+# 07. Subcription update not involving actual update
```
Comment (fgalan) [inline test/functionalTest/cases/2064_attrs_blacklist_subs/2064_attrs_blacklist_complex.test]: Subcription -> Subscription

(Maybe also needed in associated "echo" command)

File: src/lib/mongoBackend/MongoCommonSubscription.cpp (hunk 5 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonSubscription.cpp]: Yes, whitespace 'everywhere' ...

    bl? "true" : "false"

File: src/lib/mongoBackend/mongoUpdateSubscription.cpp (hunk 6 above)
Comment (crbrox) [inline src/lib/mongoBackend/mongoUpdateSubscription.cpp]: Fixed 0ec57f1

File: src/lib/mongoBackend/MongoCommonSubscription.cpp (hunk 5 above)
Comment (crbrox) [inline src/lib/mongoBackend/MongoCommonSubscription.cpp]: fixed 0ec57f1

File: src/lib/mongoBackend/mongoUpdateSubscription.cpp (hunk 7 above)
Comment (crbrox) [inline src/lib/mongoBackend/mongoUpdateSubscription.cpp]: fixed 0ec57f1

File: test/functionalTest/cases/2064_attrs_blacklist_subs/2064_attrs_blacklist_complex.test (hunk 8 above)
Comment (crbrox) [inline test/functionalTest/cases/2064_attrs_blacklist_subs/2064_attrs_blacklist_complex.test]: fixed e89fcbce739f20de09af3507da021774c9efc720

**Comment by fgalan** [comment]
LGTM

**Comment by kzangeli** [comment]
LGTM

**Pull request merged**